drivers: ethernet: phy: add support to disable autoneg

maass-hamburg opened the pull request:
Adds support for disabling Auto-negotiation.

Done by adding a flags parameter to `phy_configure_link()`.

- also improved the phy_mii in general
- put commonly used parts used in the ethernet phys into a common header, so it can be shared.

maass-hamburg committed:
drivers: ethernet: phy_mii: add support for disabling auto-neg

Adds support for disabling auto-negotiation.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
net: shell: iface: support deactivating auto-neg

support deactivating auto-neg in face shell.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: net: ethernet: phy: add support for disabling auto-negotiation

Add option in enum phy_link_speed to disable auto-negotiation.
This allows PHY drivers to support disabling auto-negotiation.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

pdgendt commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -191,20 +191,61 @@ static int update_link_state(const struct device *dev)
 	}
 
-	bmcr_reg |= MII_BMCR_AUTONEG_ENABLE | MII_BMCR_AUTONEG_RESTART;
+	if (!(bmcr_reg & MII_BMCR_AUTONEG_ENABLE)) {
```
```suggestion
	if ((bmcr_reg & MII_BMCR_AUTONEG_ENABLE) == 0U) {
```

pdgendt commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -405,87 +447,112 @@ static int phy_mii_cfg_link(const struct device *dev,
-	if (data->gigabit_supported) {
-		if (phy_mii_reg_read(dev, MII_1KTCR, &c1kt_reg) < 0) {
-			ret = -EIO;
+	if (adv_speeds & LINK_AUTO_NEGOTIATION_DISABLED) {
```
```suggestion
	if ((adv_speeds & LINK_AUTO_NEGOTIATION_DISABLED) != 0U) {
```

pdgendt commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -405,87 +447,112 @@ static int phy_mii_cfg_link(const struct device *dev,
-	} else {
-		anar_reg &= ~MII_ADVERTISE_100_HALF;
-	}
+		if (adv_speeds & LINK_FULL_10BASE) {
```
Same for the other checks

```suggestion
		if ((adv_speeds & LINK_FULL_10BASE) != 0U) {
```

maass-hamburg committed:
drivers: ethernet: phy_mii: check condition explicitly

check condition explicitly.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: ethernet: phy_mii: check return values in init

check return values in init.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented on drivers/ethernet/phy/phy_mii.c:
```diff
@@ -89,7 +89,7 @@ static inline int phy_mii_reg_write(const struct device *dev, uint16_t reg_addr,
 }
 
-static bool is_gigabit_supported(const struct device *dev)
+static int is_gigabit_supported(const struct device *dev)
```
changed

jukkar commented:
The API seems to be unstable so I think it is ok to adjust to it (perhaps updating the version number to 0.8.1). With this in mind, I think this looks better.
```
static inline int phy_configure_link(const struct device *dev, enum phy_link_speed speeds, bool use_auto_negotiation)
```

maass-hamburg committed:
drivers: ethernet: phy: remove unused/wrong cfg_link

if we can't use cfg_link to (re-)configure the link,
we don't need it and shouldn't have it.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: ethernet: mii: provide bit val

Provide bit value for control registers
to be able to use WRITE_BIT() macro.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: ethernet: phy: add common functions for PHYs

This commit adds common functions for PHYs to the mii driver.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: ethernet: phy: phy_mii: start autoneg in cfg_link

already (re-)start autonegotiation in phy_mii_cfg_link.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
doc: migration guide: ethernet: phy:  mention phy_configure_link change

mention change of phy_configure_link().

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

clamattia commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -471,7 +471,7 @@ static int nxp_enet_phy_configure(const struct device *phy, uint8_t phy_mode)
 
 	/* Configure the PHY */
-	ret = phy_configure_link(phy, speeds);
+	ret = phy_configure_link(phy, speeds, 0);
```
Sorry, nvm. The problem I was linking to was in a `header` and its `static inline`-function.
My bad.

maass-hamburg replied:
will revert it back

manuargue commented on include/zephyr/net/phy.h:
```diff
@@ -176,7 +182,8 @@ __subsystem struct ethphy_driver_api {
 
 	/** Configure link */
-	int (*cfg_link)(const struct device *dev, enum phy_link_speed adv_speeds);
+	int (*cfg_link)(const struct device *dev, enum phy_link_speed adv_speeds,
```
`adv_speeds` now has an implicit dual meaning: either a list of advertised speeds when auto-negotiation is enabled, or a forced mode when auto-negotiation is disabled. That overload is a little awkward.

What do you think about adding a check in `phy_configure_link()` that returns `-EINVAL` if auto-negotiation is disabled and more than one speed/duplex mode is specified?

maass-hamburg replied:
Sure

clamattia commented on include/zephyr/net/mii.h:
```diff
@@ -54,32 +54,41 @@
 #define MII_ESTAT      0xf
 
 /* Basic Mode Control Register (BMCR) bit definitions */
+#define MII_BMCR_RESET_BIT           15
```
The `if (IS_BIT_SET(x, MII_BMCR_RESET_BIT)) {` -pattern could be nice in combination with these. No need to rewrite anything. Just wanted to point it out, in case you were not aware.
